### DISCUSSION TITLE
Consider added functionality to word: true, user-defined keys to expand instead of <SPC> & allow user and packages to decide whether leading/trailing whitespace is required for expansion

### DISCUSSION THREAD

**Discussion by CillySu** [espanso/espanso#1412] [Feature requests and Ideas]
I have crossposted this feature request in Issues, because I feel like it is half feature request and half a problem with the current `word: true` behaviour.

I use `word: true` for a large number of my triggers, I find it much more beneficial as it allows me to have multiple expansions with the same starting characters.

With `word: true` enabled I am able to use "v" → vein and "vu" → venule, for example. This obviously would require a prepending in the absence of word: true such as ":v".

As of right now, the only way to expand `word: true` if _prepended_ **and** _appended_ by `<SPC>`,  this presents a number of issues:

1. Expansions will not expand if they are the first character in a line, as they are not prepended by a whitespace. This is problematic when using bullet points, or when not writing in long-form paragraphs, as beginning on a new line is rather common.
2. Sometimes I wish for a trigger to expand if appended by `<SPC>` but irrespective of the preceding content (or the reverse, where there is a space before but I don't care about what comes after)

For example:
```
- trigger: "dyn"
   replace: "dynamics"
```
where the intended use is for the user to type `"pharmacodyn`<SPC>`" → "pharmacodynamics"`, but to **NOT** have espanso expand when typing "dynamo", for example.

6. Triggers will not expand if they are followed by certain characters which should probably be expected as default intended behaviour, such as colon, semicolon, closing brackets, `<TAB⇥>`.
7. It does not facilitate the expansion of a trigger into a shortened form of the expansion (think of trigger "pco" → "pharmaco ") with the intention of the user typing "pco`<SPC>`dyamics" to achieve "pharmacodynamics" or "pco`<SPC>`kinetics" to achieve "pharmacokinetics" — this is not possible in `word: true` current behaviour because the expansion appends the space automatically.

I understand that no. 4 can be remedied by using:
```
- trigger: "pco"
   replace: "pharmaco$|$"
   word: true
```

and this works quite well. But I have a solution i propose to be of better utility.

My suggestion is, a **master config** for `word: true` behaviour in the user's config file as well as the ability for packages to override this behaviour (if permitted by the user) in the package.yml.

I would suggest having the ability to have a key or set of keys (I would likely use `<TAB ⇥>`) which allows me to expand the trigger without espanso injecting the whitespace after the expansion "pharmaco_" where _ is the space.

```
default.yml or package.yml
word_boundaries: [TAB, ENTER, SPACE, SHIFT, "}", "]", ">", "\"", "/", "-"]
word_defaults: [true, pre, post, append]
\# true = triggers obey word:true by default unless explicitly stated otherwise
\# pre = requires leading whitespace for expansion of trigger
\# post = requires trailing whitespace for expansion of trigger
\# append = whitespace is injected after expansion

word_override: true
\# allows the user to permit packages to override their defaults for the package's expansions

base.yml
matches:
   - trigger: "trigger"
     replace: "expansion"
     word: [true, pre, post, append]
     \# Permit the overriding of package and user defaults
```

I personally would have `word: true` expand with closing parentheses `( } / ] / > )` and quotation marks but ideally for power-users this should be user-definable.

It is possible that some of this proposed behaviour strays beyond what `word: true` was intended to do, in this case I believe a new config option should be permitted.

**Comment by CillySu** [comment]
>I understand that no. 4 can be remedied by using:

There was an issue with the numbering, no. 4 **should read no. 2** here

**Comment by kpym** [comment]
@CillySu isn't the existing [`word_separators`](https://espanso.org/docs/configuration/options/) what you propose as `word_boundaries`?

**Reply by CillySu** [reply]
No. Because under the current functionality, the boundary needs to be present both before and after the trigger in order for expansion to occur.

**Comment by kpym** [comment]
I think that some of your cases can be solved by `regex`. For example for 2 and 6:
```yaml
  - regex: "dyn\\s"
    replace: "dynamics"
  - regex: "dyn\\b(?P<end>[^\\s])"
    replace: "dynamics{{end}}"
```
Now if you type `pharmacodyn` + <kbd>any space character</kbd> you get `pharmacodynamics` (without the space character). And if you type something like `pharmacodyn,` you get `pharmacodynamics,` (the non space, non letter character is preserved, like `,` here).

Is this what you want ?

**Reply by CillySu** [reply]
It's a decent workaround yes, but it doesn't follow a list of characters such as those defined in word_separators, it's also difficult for less tech competent users (me)

Thank you for coming up with a good solution though!

**Reply by kpym** [reply]
You can adjust the list. If you want to keep the last character you ca do for example:
```yaml
  - regex: "dyn(?P<end>[\\s})</-\\\\])"
    replace: "dynamics{{end}}"
```
Here `[\\s})</-\\\\]` means "one ot the characters `space`, `tab`, `enter`, `}`,`)`... The `\\s` means "any space character" and `\\\\` means simply `\` but because `\` is a special one we have to "escape it".
